File: c++/design_pattern/Builder.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
#include <initializer_list>
#include <memory>

using namespace std;


namespace ProblemCase{

	struct HtmlElement{
		string name;
		string text;
		::std::vector<HtmlElement> elements;
		HtmlElement(): name(""),text(""),elements(0){};
		HtmlElement( string& name,  string& text) : name(name), text(text), elements(0) {}

		::std::string str(int indent = 0) const {

			::std::ostringstream oss;

			oss << "<" << name << ">" <<::std::endl;
			for(auto& it : elements){
				oss<<it.str();
			}
			oss << text <<::std::endl;
			oss<< "</" << name << ">" << ::std::endl;

			return oss.str();	
		}
	};
	
	void do_problem_case(){
		using namespace std;
		string words[2]= {"hello", "world"};
		string child_element_name = "li";
		string parent_element_name = "ul";
		string empty = "";

		::ProblemCase::HtmlElement list(parent_element_name,empty);
		for(int i = 0; i < 2; i++){
			string w = words[i];
			list.elements.emplace_back(HtmlElement(child_element_name, w));
		}
		std::cout << list.str() << endl;
	}

}

namespace SimpleBuilderCase{
	typedef ::ProblemCase::HtmlElement HtmlElement;

	struct HtmlBuilder {
		HtmlElement root;
		HtmlBuilder(string root_name){root.name = root_name;}

		void add_children(string child_name, string child_text){
			root.elements.emplace_back(HtmlElement(child_name, child_text));
		}
	};


	void do_simple_budilder_case(){
		using namespace std;
		string root_name = "ul";
		string child_name = "li";
		string words[2] = {"Hello", "World!"};
		HtmlBuilder htmlbuilder(root_name);
		for(auto &it : words){
			htmlbuilder.add_children(child_name,it);
		}
		cout << "== simple builder=="<<endl;
		cout << htmlbuilder.root.str();

	}
};


namespace FluentInterfaceBuilderCase{
	typedef::ProblemCase::HtmlElement HtmlElement;

	struct HtmlBuilder {
		HtmlElement root;
		HtmlBuilder(string root_name){root.name = root_name;}

		HtmlBuilder& add_children(string child_name, string child_text){
			root.elements.emplace_back(HtmlElement(child_name, child_text));
			return *this;
		}
	};
	struct HtmlBuilder2 {
		HtmlElement root;
		HtmlBuilder2(string root_name){root.name = root_name;}

		HtmlBuilder2* add_children(string child_name, string child_text){
			root.elements.emplace_back(HtmlElement(child_name, child_text));
			return this;
		}
	};


	void do_fluent_interface_budilder_case(){
		using namespace std;
		string root_name = "ul";
		string child_name = "li";
		string words[2] = {"Hello", "World!"};
		HtmlBuilder htmlbuilder(root_name);
		// HtmlBuilder2* htmlbuilder = new HtmlBuilder2(root_name);

		htmlbuilder.add_children(child_name, words[0]).add_children(child_name,words[1]);
		// htmlbuilder2->add_children(child_name, words[0])->add_children(child_name,words[1]);
		
		cout << "== fluent interface builder=="<<endl;
		cout << htmlbuilder.root.str();

	}

};

/*사용자에게 빌더 클래스를 사용해야한단는 것을 알리는 방법*/
namespace HowToNotifyUsersToUseBuilderClass{

	typedef ::ProblemCase::HtmlElement HtmlElement;
	// typedef ::FluentInterfaceBuilderCase::HtmlBuilder HtmlBuilder;

	struct HtmlBuilder : ::FluentInterfaceBuilderCase::HtmlElement{

		HtmlElement root;
		HtmlBuilder(string root_name){root.name = root_name;}

		/* 3. 빌더가 아닌 원래 클래스를 생성할 수 있도록 하기 */
		HtmlBuilder& add_children(string child_name, string child_text){
			root.elements.emplace_back(HtmlElement(child_name, child_text));
			return *this;
		}
		operator HtmlElement() const {return root;}
	};

	struct EnhancedHtmlElement : HtmlElement{
		string name;
		string text;
		/* 1. 팩터리 메서드르를 두어 빌더를 생성할 수 있도록 하기 */
		static unique_ptr<HtmlBuilder> build(const string& root_name){
			return make_unique<HtmlBuilder>(root_name);
		}

		/* 2. 모든 생성자를 숨기기 */
		protected : 
			EnhancedHtmlElement() {}
			EnhancedHtmlElement(const string& name, const string& text) : name(name), text(text) {};
	};

	void how_to_notify_uses_to_use_builder_class(){
		using namespace std;
		string root_name = "ul";
		string child_name = "li";
		string words[2] = {"Hello", "World!"};
		auto htmlbuilder = EnhancedHtmlElement::build(root_name);
		HtmlElement e = htmlbuilder->add_children(child_name,words[0]).add_children(child_name,words[1]);
		cout << "== how_to_notify_uses_to_use_builder_class =="<<endl;
		// cout << htmlbuilder->root.str();
		cout << e.str() << endl;

	}
}



namespace groovy_style{
	struct Tag {
		std::string name;
		std::string text;
		std::vector<Tag> children;
		std::vector<std::pair<std::string,std::string>> attributes;

		friend std::ostream& operator<<(std::ostream& os, const Tag& tag){
			
			if(!tag.attributes.empty()){
				os << "<" << tag.name<< " ";
				for (auto& it : tag.attributes){
					os << it.first <<"="<<it.second <<"";
				}
				os << ">" << std::endl;

			}else{
				os << "<" << tag.name <<">" << std::endl;
			}
			os << tag.text << std::endl;
			for(auto& it : tag.children){
				 os << it;
			}
			os << "</" << tag.name << ">" << std::endl;
			return os;
		};


		protected : 
			Tag(const std::string& name, const std::string& text):name(name), text(text){};
			Tag(const std::string& name, const std::vector<Tag>& child) : name(name), children(child){};

	};

	struct P : Tag {
		std::string this_name = "p";
		explicit P(const std::string& text) : Tag("p",text){}
		P(std::initializer_list<Tag> children) : Tag("p",children){}
	};


	struct IMG : Tag{
		explicit IMG(const std::string& url) : Tag("img",""){
			attributes.emplace_back(make_pair("src",url));
		};
	};


	void do_groovy_style(){
		cout << "== groovy_style =="<<std::endl;

		std::cout << 
		P{
			IMG{"http://url"}
		} << std::endl;
	}
}


namespace compositeBuilderCase{
	class PersonBuilder;
	class PersonAddressBuilder;
	class PersonJobBuilder;
	class Person{

		std::string street_address, post_code,city;
		std:: string company_name, position;
		int annual_income = 0;

		Person(){}

		public : 
			static PersonBuilder create();

			Person(Person&& other) 
				: street_address(other.street_address),
				post_code(other.post_code),
				city(other.city),
				company_name(other.company_name),
				position(other.position),
				annual_income(other.annual_income){};

		friend class PersonBuilder;
		friend class PersonAddressBuilder;
		friend class PersonJobBuilder;
	};

	class PersonBuilderBase{
		protected:
			Person& person;
			explicit PersonBuilderBase(Person& person) : person(person){}
		public : 
			operator Person() const{
				return std::move(person);
			}
			PersonAddressBuilder lives() const;
			PersonJobBuilder works() const;
	};
	class PersonAddressBuilder : public PersonBuilderBase{
		typedef PersonAddressBuilder Self;

		public :
			explicit PersonAddressBuilder(Person& person) : PersonBuilderBase(person) {}

			Self& at(std::string street_address){
				person.street_address = street_address;
				return *this;
			}
			Self& with_postcode(std::string post_code){
				person.post_code = post_code;
				return *this;
			}
			Self& in(std::string city){
				person.city = city;
				return *this;
			}
	};
	class PersonJobBuilder : public PersonBuilderBase{
		typedef PersonJobBuilder Self;

		public : 
			explicit PersonJobBuilder(Person& person) : PersonBuilderBase(person){}

			Self& at(std::string company_name){
				person.company_name = company_name;
				return *this;
			};
			Self& as_a(std::string position){
				person.position = position;
				return *this;
			};
			Self& earning(int annual_income){
				person.annual_income = annual_income;
				return *this;
			};
	};
	class PersonBuilder : public PersonBuilderBase{
			Person p;
		public :
			PersonBuilder() : PersonBuilderBase(p){}

	};


	PersonBuilder Person::create(){
		return PersonBuilder();
	}

	PersonAddressBuilder PersonBuilderBase::lives() const {
		return PersonAddressBuilder(person);
	};

	PersonJobBuilder PersonBuilderBase::works() const{
		return PersonJobBuilder(person);
	}



	void do_composite_builder_case(){
		cout << "== composite_builder_case =="<<std::endl;

		Person p = Person::create()
			.lives().at("123 London Road")
				.with_postcode("SW1 1GB")
				.in("London")
			.works().at("PragmaSoft")
				.as_a("Consultant")
				.earning(10e6);
	}


};



int main(void){

	::ProblemCase::do_problem_case();
	::SimpleBuilderCase::do_simple_budilder_case();
	::FluentInterfaceBuilderCase::do_fluent_interface_budilder_case();
	::groovy_style::do_groovy_style();
	::compositeBuilderCase::do_composite_builder_case();

	return 0;
}

File: java/src/main/java/com/example/design_pattern/Builder.java
package com.example.design_pattern;


import java.util.*;

public class Builder {

	private static class FluentInterfaceBuilder{
		
		private static class HtmlElement{
			private String name;
			private String text;
			private List<HtmlElement> children;


			public HtmlElement(String name, String text){
				this.name = name;
				this.text = text;
				this.children = new ArrayList<>();
			}

			public HtmlElement(){
				this("","");
			}

			public void addChildren(String name, String text){
				children.add(new HtmlElement(name,text));
			}

			@Override
			public String toString(){
				StringBuilder builder = new StringBuilder();
				builder.append(String.format("<%s>\n",name));
				for(HtmlElement child : children){
					builder.append(child);
				}
				builder.append(text+"\n");
				if(!"".equals(text)){
					builder.append(String.format("</%s>\n",name));
				}
				return builder.toString();
			}

			public void setName(String name){
				this.name = name;
			}

			public void SetText(String text){
				this.text = text;
			}

		}

		private static class HtmlElementBuilder{
			private HtmlElement root;

			public HtmlElementBuilder(String rootName, String rootText){
				this.root = new HtmlElement(rootName, rootText);
			}

			public HtmlElementBuilder(String rootName){
				this.root = new HtmlElement(rootName, "");
			}

			public HtmlElementBuilder addChildren(String name, String text){
				root.addChildren(name, text);
				return this;
			}

			public HtmlElement build(){
				return root;
			}
		}

		public static void main(String... args){
			HtmlElementBuilder builder = new HtmlElementBuilder("ul");
			HtmlElement root = builder.addChildren("li", "Hello")
									.addChildren("l1", "World")
									.build();

			System.out.println(root.toString());
		}
	}

	private static class CompositeBuilder {

		private static class Person {
			private String streetAddress;
			private String postCode;
			private String city;
			private String companyName;
			private String position;
			private int annualIncome;

			public Person(){}

			public Person(Person other){
				this.streetAddress = other.streetAddress;
				this.postCode = other.postCode;
				this.city = other.city;
				this.companyName = other.companyName;
				this.position = other.position;
				this.annualIncome = other.annualIncome;
			}

			void setStreetAddress(String streetAddress){
				this.streetAddress = streetAddress;
			}

			void setPostCode(String postCode){
				this.postCode = postCode;
			}

			void setCity(String city){
				this.city = city;
			}

			void setCompanyName(String companyName){
				this.companyName = companyName;
			}

			void setPosition(String position){
				this.position = position;
			}

			void setAnnualIncome(int annualIncome){
				this.annualIncome = annualIncome;
			}
		}

		private static class PersonBuilder{

			private Person p;

			public PersonBuilder(){
				this.p = new Person();
			}

			public Person build(){
				return p;
			}

			public PersonAddressBuilder lives(){
				return new PersonAddressBuilder(p);
			}

			public PersonJobBuilder works(){
				return new PersonJobBuilder(p);
			}
		}

		private static class PersonAddressBuilder extends PersonBuilder{
			private Person p;
			public PersonAddressBuilder(Person p){
				this.p = p;
			}

			public PersonAddressBuilder at(String streetAddress){
				p.setStreetAddress(streetAddress);
				return this;
			}

			public PersonAddressBuilder withPostCode(String postCode){
				p.setPostCode(postCode);
				return this;
			}

			public PersonAddressBuilder in(String city){
				p.setCity(city);
				return this;
			}
		}

		private static class PersonJobBuilder extends PersonBuilder{
			private Person p;
			public PersonJobBuilder(Person p){
				this.p = p;
			}

			public PersonJobBuilder at(String companyName){
				p.setCompanyName(companyName);
				return this;
			}

			public PersonJobBuilder asA(String position){
				p.setPosition(position);
				return this;
			}

			public PersonJobBuilder earning(int annualIncome){
				p.setAnnualIncome(annualIncome);
				return this;
			}
		}

		public static void main(String... args) {
			PersonBuilder builder = new PersonBuilder();

			Person p = builder.lives()
							.at("123 London Road")
							.withPostCode("SW1 1GB")
							.in("London")
							.works()
							.at("PragmaSoft")
							.asA("Consultant")
							.earning((int)10e6)
							.build();
		}
	}

	private static class HierarchicalBuilder {


		private abstract class Pizza{
			public enum Topping {HAM, MUSHROOOM, ONION, PEPPER, SAUSAGE};
			private Set<Topping> toppings;
			abstract static class Builder<T extends Builder<T>>{
				Enum<Set<Topping> toppings = EnumSet.noneOf(Topping.class);

				public T addTopping(Topping topping){
					toppings.add(Object.requireNonNull(topping));
					return self();
				}

				abstract Pizza build();
				protected abstract T self();
			}
		}

		private class NyPizza extends Pizza {
			public enum Size {SMALL, MEDIUM, LARGE};
			private Size size;

			public static class Builder extends Pizza.Builder<Builder> {
				private final Size size;

				public Builder(size size){
					this.size= Objects.requireNonNull(size);
				}

				@Override
				public NyPizze build(){
					return new NyPizze(this);
				}

				@Override
				protected Bilder self() {return this;}
			}

			private NyPizze(Builder builder){
				super(builder);
				size = builder.size();
			}
		}

		private class Calzone extends Pizza{
			private boolean sauceInside;

			public static class Builder extends Pizza.Builder<Builder> {
				private boolean sauceInside = false;

				public Builder sauceInside() {
					sauceInside = true;
					return this;
				}

				@Override 
				public Calzone build(){
					return new Calzone(this);
				}

				@Override
				protected Builder self() {
					return this;
				}
			}

			private Calzone(Builder builder){
				super(builder);
				sauceInside = builder.sauceInside;
			}	
		}

		public static void main(String... args){
			NyPizze pizza = new NyPizze.Builder(SMALL)
				.addTopping(SAUSAGE). addTopping(ONION).build();
			Calzone calzone = new Calzone.Builder()
				.addTopping(HAM).sauceInside().build();
		}
	}

	public static void main(String... args){
		FluentInterfaceBuilder.main();
		CompositeBuilder.main();
		HierarchicalBuilder.main();
	}
}